File: ruby/crypto-square/crypto_square.rb
class Crypto
  def ciphertext
    return '' if @columns.zero?

    chuncked_plaintext.transpose.map(&:join).join(' ')
  end

  private

  attr_reader :plaintext, :columns

  def initialize(plaintext)
    @plaintext = plaintext.downcase.gsub(/[^\w]/, '')
    @columns = Math.sqrt(@plaintext.length).ceil
  end

  def chuncked_plaintext
    chunks = plaintext.chars.each_slice(columns).to_a
    chunks.last << ' ' while chunks.last.length < columns
    chunks
  end
end

File: clojure/crypto-square/src/crypto_square.clj
(ns crypto-square
  (:require [clojure.string :as str]))

(defn- pad-right [s to-size]
  (apply str s (repeat (- to-size (count s)) \space)))

(defn- transpose [m]
  (apply map (comp str/join list) m))

(defn rotate-segments [segments]
  (let [max-size (apply max (map count segments))]
    (->> segments
         (map #(pad-right % max-size))
         transpose
         (map str/trim))))

(defn normalize-plaintext [plaintext]
  (-> plaintext
      str/lower-case
      (str/replace #"[^a-z0-9]*" "")))

(defn square-size [text]
  (int (Math/ceil (Math/sqrt (count text)))))

(defn plaintext-segments [text]
  (let [text (normalize-plaintext text)
        size (square-size text)]
    (map str/join (partition-all size text))))

(defn normalize-ciphertext [text]
  (->> text
       plaintext-segments
       rotate-segments
       (str/join " ")))

(defn ciphertext [text]
  (str/replace (normalize-ciphertext text) " " ""))
